lib/install/validate/validator.go

```diff
@@ -337,6 +340,74 @@ func (v *Validator) ValidateTarget(ctx context.Context, input *data.Data, allowE
 	return conf, v.ListIssues(op)
 }
 
+func (v *Validator) ValidateStorageQuota(ctx context.Context, quotaGB int, conf *config.VirtualContainerHostConfigSpec, calUsage bool) error {
+	op := trace.FromContext(ctx, "ValidateStorageQuota")
+	defer trace.End(trace.Begin("", op))
+	op.Info("Validating storage quota")
+
+	vmStorageUsage := int64(0)
+	imageStorageUsage := int64(0)
+	var err error
+	if calUsage {
+		vmStorageUsage, err = v.getVMStorageUsage(op, conf)
+		if err != nil {
+			return err
+		}
+		imageStorageUsage, err = v.getImageStorageUsage(op, conf)
+		if err != nil {
+			return err
+		}
+	}
+	if int64(quotaGB)*units.GiB > v.getDatastoreFreeSpace(op)+vmStorageUsage+imageStorageUsage {
+		return errors.New("Storage quota exceeds datastore free space")
+	}
+	return nil
+}
+
+func (v *Validator) getVMStorageUsage(op trace.Operation, conf *config.VirtualContainerHostConfigSpec) (int64, error) {
+	imageURL := conf.ImageStores[0]
+	// get a ds helper for this ds url
+	dsHelper, err := datastore.NewHelper(trace.NewOperation(op, "datastore helper creation"), v.session,
+		v.session.Datastore, fmt.Sprintf("%s/%s", imageURL.Path, constants.KVStoreFolder))
+	if err != nil {
+		return 0, err
+	}
+
+	kv, err := kvstore.NewKeyValueStore(op, kvstore.NewDatastoreBackend(dsHelper), APIKV)
+	if err != nil {
+		return 0, err
+	}
+
+	if kv != nil {
+		storageUsageBytes, err := kv.Get("docker.storage.usage")
```

Comment:
I'm surprised we're using the kvstore for this instead of the logic for calculating from the datastore. While it's unusual for a viadmin to manually delete containers via the vSphere API it can be done (you have to enable the destroy method first) and I'd expect the change to be reflected here. What was the reasoning for using a derived value?

Reply:
This solution does not consider the operations out of VCH controls, because if attackers have control of this vCenter, they have many ways to break VCH or use up datastores regardless storage quota settings. Thus, a value kept in kvstore will make calculation simpler and straightforwards.

Reply:
This concern isn't about an attack; an authentic vSphere Administrator could delete containers via the vSphere API. Being able to treat container VMs like VMs is a selling point for VIC.

Reply:
We have the destroy method disabled for containerVMs via the vSphere API only because the method is lacking support for controlling the associated disk deletion behaviours. There should be (I couldn't find it quickly) either a Troubleshooting section or a knowledge base article about how to delete containerVMs via the vSphere API when the Docker API is unavailable.

While this is an edge case consideration for cVMs and images, it's a much more likely proposition with volumes/volume stores as they are directly visible and easy to delete via the UI. While the various volume store comments are forward looking, it makes sense to write code that can be re-used where possible.

We already have code that does the datastore search and calculation - is that not comparatively simple to call here?

Reply:
I tried to calculate the container vm file sizes through vm/datastore traverse in the latest fixuped codes. The function is correct, and the time to traverse is longer. Considering configure is an one-time operation, we can accept the slowing down.